### ISSUE TITLE
Compiler doesn't compile when using exactOptionalPropertyTypes regression

### ISSUE THREAD

**Issue by BenjaminHutchinson** [angular/angular#46148]
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

Yes

### Description

Hi,

I am currently experiencing an issue while trying to use the Typescript flag 'exactOptionalPropertyTypes'.

Compilation is failing due to some mitmatched types around the angular compiler.

It appears to be the same issue as: https://github.com/angular/angular/issues/44157 which should be fixed in the version of Angular i'm running (13.3.5). 

I believe the fix would have been included in 13.1 but I could not find anything mentioned about it in the changelog?

### Please provide the exception or error you saw

```true
------------------------------------------------------------------------------
Building entry point '@horus/core/rx'
------------------------------------------------------------------------------
× Compiling with Angular sources in Ivy full compilation mode.
node_modules/@angular/compiler/src/render3/view/template.d.ts:220:22 - error TS2420: Class 'BindingScope' incorrectly implements interface 'LocalResolver'.
  Types of property 'globals' are incompatible.
    Type 'Set<string> | undefined' is not assignable to type 'Set<string>'.
      Type 'undefined' is not assignable to type 'Set<string>'.

220 export declare class BindingScope implements LocalResolver {
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
OS: win32 x64

Angular: 13.3.5
... animations, cdk, common, compiler, compiler-cli, core, forms
... language-service, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1303.4
@angular-devkit/build-angular   13.3.4
@angular-devkit/core            13.3.4
@angular-devkit/schematics      13.3.4
@angular/cli                    13.3.4
@schematics/angular             13.3.4
ng-packagr                      13.3.1
rxjs                            7.5.5
typescript                      4.6.4
```

**Comment by JoostK** [comment]
If this does indeed happen then it appears to be a TypeScript bug, as the types are identical in the source; only the `.d.ts` file as emitted by TypeScript introduces the slight difference in type. This should have been resolved in TypeScript 4.5 (which became supported in 13.1) and your version information indicates that 4.6 is being used, so either this was a TS regression _or_ some older version of TypeScript is being used here.

Could you please provide a reproduction of this issue so we can investigate further? Also note that v14 should not be affected as the typings are now being bundled, meaning that internal types are no longer part of the public `.d.ts` file.

**Comment by BenjaminHutchinson** [comment]
I believe I've found the source of the error, it happens when including the ngx-logger library, I believe they're still publishing for an old version of angular.

https://stackblitz.com/edit/node-kmwkew
